File: day2/cxx/puzzle1.cxx
#include <algorithm>
#include <fstream>
#include <iostream>
#include <unordered_map>
#include <string>

#include "lines.hpp"

using namespace std;

static unordered_map<string, uint32_t> win_lose {
    {"A X", 3}, {"A Y", 6}, {"A Z", 0},
    {"B X", 0}, {"B Y", 3}, {"B Z", 6},
    {"C X", 6}, {"C Y", 0}, {"C Z", 3},
};

const auto solve()
{
    ifstream f{"../input.txt"};

    auto score = 0u;

    for (auto & line: lines(f))
    {
        auto choice = (line[2] - 'X' + 1);
        auto outcome = win_lose[line.substr(0,3)]; 
        score += choice + outcome;
    }

    return score;
}

int main()
{
    auto solution = solve();
    cout << solution << endl;
}

File: day2/py/puzzle1.py
#!/usr/bin/env python

XYZ = { 'X': 1, 'Y': 2, 'Z': 3}

win_lose = {
    'X': { 'A': 3, 'B': 0, 'C': 6 },
    'Y': { 'A': 6, 'B': 3, 'C': 0 },
    'Z': { 'A': 0, 'B': 6, 'C': 3 }
}

def solve():
    score = 0

    with open('../input.txt', 'r') as f:
        for line in f:
            opponent = line[0]
            player = line[2]

            score += XYZ[player] + win_lose[player][opponent]        

    return score

if __name__ == "__main__":
    solution = solve()
    print(solution, )

